### PR TITLE
[SPARK-34541][CORE] Fixed an issue where data could not be cleaned up when unregisterShuffle.

### PR DESCRIPTION
**Pull request by yikf**
### What changes were proposed in this pull request?
Fixed an issue where data could not be cleaned up when unregisterShuffle.

### Why are the changes needed?
While we use the old shuffle fetch protocol, we use partitionId as mapId in the ShuffleBlockId construction,but we use `context.taskAttemptId()` as mapId that it is cached in `taskIdMapsForShuffle` when we `getWriter[K, V]`.

where data could not be cleaned up when unregisterShuffle ,because we remove a shuffle's metadata from the `taskIdMapsForShuffle`'s mapIds, the mapId is `context.taskAttemptId()` instead of partitionId.

### Does this PR introduce _any_ user-facing change?
yes

### How was this patch tested?
exist test.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala [hunk 1]
```diff
@@ -28,7 +28,7 @@ class StringExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
   test("concat") {
     def testConcat(inputs: String*): Unit = {
       val expected = if (inputs.contains(null)) null else inputs.mkString
-      checkEvaluation(Concat(inputs.map(Literal.create(_, StringType))), expected, EmptyRow)
+      checkEvaluation(Concat(inputs.map(Literal.create(_, StringType))), expected)
```
Comment (yikf) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala:31]: For reviewer : clean up unused code for code simplifications.
Method signature for checkEvaluation as follow:
checkEvaluation(expression: => Expression, expected: Any, inputRow: InternalRow = EmptyRow)

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleManager.scala [hunk 2]
```diff
@@ -154,7 +154,7 @@ private[spark] class SortShuffleManager(conf: SparkConf) extends ShuffleManager
       metrics: ShuffleWriteMetricsReporter): ShuffleWriter[K, V] = {
     val mapTaskIds = taskIdMapsForShuffle.computeIfAbsent(
       handle.shuffleId, _ => new OpenHashSet[Long](16))
-    mapTaskIds.synchronized { mapTaskIds.add(context.taskAttemptId()) }
```
Comment (yikf) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleManager.scala:157]: For more convenient review:

ShuffleMapTask#runTask
`val mapId = if (SparkEnv.get.conf.get(config.SHUFFLE_USE_OLD_FETCH_PROTOCOL)) {
      partitionId
    } else context.taskAttemptId()`

SortShuffleManager#getWriter
`val mapTaskIds = taskIdMapsForShuffle.computeIfAbsent(
      handle.shuffleId, _ => new OpenHashSet[Long](16))
    mapTaskIds.synchronized { mapTaskIds.add(context.taskAttemptId()) }`

SortShuffleManager#unregisterShuffle
`Option(taskIdMapsForShuffle.remove(shuffleId)).foreach { mapTaskIds =>
      mapTaskIds.iterator.foreach { mapTaskId =>
        shuffleBlockResolver.removeDataByMap(shuffleId, mapTaskId)
      }
    }`

**Comment by yikf** [comment]
gentle ping @srowen , thanks for taking a look.

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleManager.scala (hunk 2 above)
Comment (srowen) [reply core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleManager.scala:157]: @xuanyuanking do you have an opinion here? I think you wrote this piece of the code.

**Review by srowen — COMMENTED** [review]
This is open vs branch-3.0 - needs to be master

**Comment by mridulm** [comment]
The existing tests are not catching this issue - can you add something to ensure we test for this problem ?

**Comment by mridulm** [comment]
+CC @otterc

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala (hunk 1 above)
Comment (otterc) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala:31]: Is this test change related to the code change?

**Review by otterc — COMMENTED** [review]
The change in `SortShuffleManager` looks good to me. However, we should add a UT to cover this. Also, I don't quite get how the change in `StringExpressionsSuite.scala` is related to this bug fix.

**Comment by yikf** [comment]
> The change in `SortShuffleManager` looks good to me. However, we should add a UT to cover this. Also, I don't quite get how the change in `StringExpressionsSuite.scala` is related to this bug fix.

The change in  `StringExpressionsSuite.scala` is unrelated to this bug fix, for code simplifications only.

**Comment by yikf** [comment]
> The existing tests are not catching this issue - can you add something to ensure we test for this problem ?

i will add the UT for this problem.

**Comment by HyukjinKwon** [comment]
@yikf, can you open a PR against master branch?

**Comment by yikf** [comment]
> @yikf, can you open a PR against master branch?

ok, should i close the currently PR?

**Comment by yikf** [comment]
open a PR [https://github.com/apache/spark/pull/31664](url) against master branch

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleManager.scala (hunk 2 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleManager.scala:157]: Thanks for pinging me, I'm reviewing #31664

**Pull request closed without merging**